core/src/sigverifier/bls_sigverifier.rs

```diff
@@ -142,6 +227,218 @@ impl BLSSigVerifier {
+            }
+            solana_signer_store::Decoded::Base3(bit_vec1, bit_vec2) => {
+                self.verify_base3_certificate(cert_to_verify, &bit_vec1, &bit_vec2, key_to_rank_map)
+            }
+        }
+    }
+
+    fn verify_base2_certificate(
+        &self,
+        cert_to_verify: &CertToVerify,
+        bit_vec: &BitVec<u8, Lsb0>,
+        key_to_rank_map: &Arc<BLSPubkeyToRankMap>,
+    ) -> bool {
+        let Some(original_vote) =
+            certificate_to_vote_message_base2(&cert_to_verify.cert_message.certificate)
+        else {
+            return false;
+        };
+
+        let Ok(signed_payload) = bincode::serialize(&original_vote) else {
+            return false;
+        };
+
+        let Some(aggregate_bls_pubkey) = aggregate_keys_from_bitmap(bit_vec, key_to_rank_map)
+        else {
+            return false;
+        };
+
+        aggregate_bls_pubkey
+            .verify_signature(&cert_to_verify.cert_message.signature, &signed_payload)
+            .unwrap_or(false)
+    }
+
+    fn verify_base3_certificate(
+        &self,
+        cert_to_verify: &CertToVerify,
+        bit_vec1: &BitVec<u8, Lsb0>,
+        bit_vec2: &BitVec<u8, Lsb0>,
+        key_to_rank_map: &Arc<BLSPubkeyToRankMap>,
+    ) -> bool {
+        let Some((vote1, vote2)) =
+            certificate_to_vote_messages_base3(&cert_to_verify.cert_message.certificate)
+        else {
+            return false;
+        };
+
+        let Ok(signed_payload1) = bincode::serialize(&vote1) else {
+            return false;
+        };
+        let Ok(signed_payload2) = bincode::serialize(&vote2) else {
+            return false;
+        };
+
+        let messages_to_verify: Vec<&[u8]> = vec![&signed_payload1, &signed_payload2];
+
+        // 1. Aggregate the two sets of public keys separately from the two bitmaps.
+        let Some(agg_pk1) = aggregate_keys_from_bitmap(bit_vec1, key_to_rank_map) else {
+            return false;
+        };
+        let Some(agg_pk2) = aggregate_keys_from_bitmap(bit_vec2, key_to_rank_map) else {
+            return false;
+        };
+
+        let pubkeys_affine: Vec<BlsPubkey> = vec![agg_pk1.into(), agg_pk2.into()];
+        let pubkey_refs: Vec<&BlsPubkey> = pubkeys_affine.iter().collect();
+
+        SignatureProjective::par_verify_distinct_aggregated(
+            &pubkey_refs,
+            &cert_to_verify.cert_message.signature,
+            &messages_to_verify,
+        )
+        .unwrap_or(false)
+    }
+}
+
+#[derive(Debug)]
+struct VoteToVerify<'a> {
+    vote_message: VoteMessage,
+    bls_pubkey: &'a BlsPubkey,
+    packet: PacketRefMut<'a>,
+}
+
+impl VoteToVerify<'_> {
+    fn payload_slice(&self) -> &[u8] {
+        // The bincode serialization of `ConsensusMessage::Vote(VoteMessage)` has the layout:
+        // [ 4-byte Enum Discriminant | Vote Payload | 96-byte Signature | 2-byte Rank ]
+        const BINCODE_ENUM_DISCRIMINANT_SIZE: usize = 4;
```

Comment:
Err, this makes assumptions on payload layout so it seems a bit brittle. Are you trying to avoid another VoteMessage serialize here? We can cache serialized vote if needed.

Reply:
Yes, I originally had the serialize logic here and I thought it was a little wasteful to deserialize (to distinguish vote and certificate) and then serialize again. By caching serialized vote, what do you precisely mean?

Reply:
I mean, it's very likely the same kind of vote all pour in at the same time, for example:
Notarize(5, hash1)
Finalize(6)

So we can keep a cache if serialization turns out to waste a lot of time. But let's put this in first then improve performance.

Reply:
Okay yeah that makes sense. Let me add a todo.